designs/local-pv/lvm/snapshot-restore.md

```diff
@@ -0,0 +1,163 @@
+    - On the other hand, the Volume creation request might have a data source as either a Snapshot or Persistent Volume.
+    - If data source is snapshot then create volume with datasource as snapshot.
+    - Validate request whether snapshot is thin or thick , if snapshot is thick then return not supported error.
+    - Validate LVM vg , owner node and capacity of CreateVolume request.
+- OpenEBS LVM driver updates openEBS lvm volume CR with snapshot name, node name, vg name, capacity, thinProvision values.
+- **lvm-node** on the target node finds the earlier created restore volume lvmvolume CR with snapshot as source.
+- **lvm-node** sends a volume create request to create a thin-snapshot from the thin-snapshot
+    - Backend (LVM) could internally:
+        - Creates thin snapshot of snapshot
+            - ```lvcreate -s -n openebs-lvm-restore-lv lvmvg/openebs-lvm-thin-snapshot```
+        - Activates above snapshot with Read-Write permissions which will be restored LV
+            - ```lvchange -kn -ay lvmvg/openebs-lvm-restore-lv```
+- **lvm-node** updates the status of restored LVM Volume CR
+- **lvm-controller** finds the updated status of restored LVM Volume CR
+- **lvm-controller** sends the success (or failure) to provisioner sidecar.
+
+### OpenEBS lvmsnapshot CR
+```yaml
+apiVersion: local.openebs.io/v1alpha1
+kind: LVMSnapshot
+metadata:
+  name: snapshot-0e1967eb-471d-4a7e-93b1-ba92d53c7a2a # lvm snapshot cr name
+  namespace: openebs
+spec:
+  ownerNodeID: node-0-397958
+  # origin volume capacity at the time of snapshot
+  sourceVolumeSize: "2147483648"
+  # 'thinSnapshot' specifies type of snapshot, this field will be yes 
+  # if snapshot is of thin-provisoned volume else it will be no
+  thinSnapshot: "yes"
+  volGroup: lvmvg
+status:
+  state: Ready
+```
+
+### Restore from Snapshot
+***restore-pvc.yaml***
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: pvs-restore-from-snapshot
+spec:
+  storageClassName: restore-sc
+  dataSource:
+    name: lvm-thin-snapshot # name of the thin snapshot
+    kind: VolumeSnapshot
+    apiGroup: snapshot.storage.k8s.io
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 2Gi # capacity of restored PVC
+```
+
+### OpeneEBS lvmvolume CR for restore
+```yaml
+apiVersion: local.openebs.io/v1alpha1
+kind: LVMVolume
+metadata:
+  name: pvc-97e44bbf-e916-4fe4-bca0-04e5aabcba4e
+  namespace: openebs
+spec:
+  capacity: "2147483648"
+  ownerNodeID: node-0-397958
+  source: snapshot-0e1967eb-471d-4a7e-93b1-ba92d53c7a2a # lvm snapshot CR name
+  volGroup: lvmvg
+status:
+  state: Ready
+```
+
+### Risks and Mitigations
+- Enable **thin_pool_autoextend_threshold** and **thin_pool_autoextend_percent** in **lvm.conf**. Prevents write failures when thin pool fills. Ensure Volume Group (VG) has spare capacity, otherwise auto-extend won’t help.
+
+
+## Testing
+- Restore from a thin snapshot.  
+- Restore from a thick snapshot.  
```

Comment:
You mean as a negative test? ie should fail

Reply:
yes, i will update it